File: leetcode/problems/350/350.py
class Solution(object):
    def intersect(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        shorter = 0
        if(len(nums1) < len(nums2)):
            shorter = 1
        else:
            shorter = 2
        inter = []
        ht= dict();
        if(shorter == 1):
            for i in range(0, len(nums2)):
                if nums2[i] not in ht:
                    ht[nums2[i]] = 1
                else:
                    ht[nums2[i]] = ht[nums2[i]] + 1
            for i in range(0, len(nums1)):
                if nums1[i] in ht and ht[nums1[i]] > 0:
                    inter.append(nums1[i])
                    ht[nums1[i]] = ht[nums1[i]] - 1
        else:
            for i in range(0, len(nums1)):
                if nums1[i] not in ht:
                    ht[nums1[i]] = 1
                else:
                    ht[nums1[i]] = ht[nums1[i]] + 1
            for i in range(0, len(nums2)):
                if nums2[i] in ht and ht[nums2[i]] > 0:
                    inter.append(nums2[i])
                    ht[nums2[i]] = ht[nums2[i]] - 1
        return inter

File: leetcode/problems/350/350.cc
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        vector<int> both;
        map<int, int> count1;
        for(int i = 0; i < nums1.size(); i++){
            if(count1.count(nums1[i]) > 0){
                count1[nums1[i]]++;
            } else {
                count1[nums1[i]] = 1;
            }
        }
        map<int, int> count2;
        for(int i = 0; i < nums2.size(); i++){
            if(count2.count(nums2[i]) > 0){
                count2[nums2[i]]++;
            } else {
                count2[nums2[i]] = 1;
            }
        }
        // then go through each key value and try to get the lowest number for each respective key
        map<int, int> lowest;
        for(auto pair : count1){
            if(count2.count(pair.first) > 0){
                // then pick minimum
                lowest[pair.first] = min(count1[pair.first], count2[pair.first]);
            }
        }
        for(auto pair : lowest){
            for(int i = 0; i < pair.second; i++){
                both.push_back(pair.first);
            }
        }
        return both;
    }
};